### PR TITLE
Fix Video Unavailable problem at YouTube preview

### PR DESCRIPTION
**Pull request by Sn-Kinos**
#### Summary

Fixed the problem that YouTube preview shows Video Unavailable instead of the video.
- Added the properties on the iframe for YouTube preview at `youtube_video.tsx`

#### Reproduce
Create a post that occurs Video Unavailable problem. I used https://youtu.be/edsx_MOhVnk on the below screenshots.

#### Ticket Link

No Tickets. (less than 20 lines of code change)

#### Screenshots

I haven't change the UI. But I can show about it w/ screenshots.

|  before  |  after  |
|----|----|
| image | image |

#### Release Note

```release-note
Fixed the problem that YouTube preview shows Video Unavailable instead of the video.
```

### CODE REVIEW

**Comment by pvev** [comment]
hey @Sn-Kinos , thank you for your contribution. Could you please describe the exact steps on how to reproduce the issue you are trying to solve so we can create a ticket and associate it with this fix? Thanks!

**Comment by Sn-Kinos** [comment]
> hey @Sn-Kinos , thank you for your contribution. Could you please describe the exact steps on how to reproduce the issue you are trying to solve so we can create a ticket and associate it with this fix? Thanks!

Thanks @pvev! I added the reproduce section on the text :D

**Comment by Sn-Kinos** [comment]
@pvev I wonder if there's anything else I need to do 🏃‍♂️

**Comment by pvev** [comment]
> @pvev I wonder if there's anything else I need to do 🏃‍♂️

Thanks @Sn-Kinos we will start the dev review and security review and QE process now. We will keep you posted.

File: webapp/channels/src/components/youtube_video/youtube_video.tsx
Comment (hmhealey) [inline webapp/channels/src/components/youtube_video/youtube_video.tsx:79]: Do we need to allow all these things? `picture-in-picture` and `encrypted-media` are the only ones that seem like they would be applicable to a video in MM. `autoplay` is technically applicable, but I don't think we want autoplaying video in MM

**Review by hmhealey — DISMISSED** [review]
Thanks for the PR, @Sn-Kinos! I always assumed that the fact that YouTube videos didn't play for me in MM was something on my end, not that we were missing some things on the `iframe`. Great find!

File: webapp/channels/src/components/youtube_video/youtube_video.tsx
Comment (Sn-Kinos) [reply webapp/channels/src/components/youtube_video/youtube_video.tsx:79]: It's necessary and the feature has been on `src` as `autoplay=1`.
| Without autoplay | With autoplay |
| --- | --- |
| <video src="https://github.com/mattermost/mattermost/assets/26380261/f7d1ea07-7a62-47d9-867d-23851b2232d1"> | <video src="https://github.com/mattermost/mattermost/assets/26380261/9d3428b8-eee8-4ad3-99bb-a32e51dded73"> |

Comment (hmhealey) [reply webapp/channels/src/components/youtube_video/youtube_video.tsx:79]: Oh right. I forgot we went from showing a static image to the actual video. That makes sense.

I guess web-share/clipboard-write are needed for a share button and accelerometer/gyroscope are needed for rotating the video then

**Comment by pvev** [comment]
@Sn-Kinos sorry that this review is taking longer that normal, but there are some security concerns about this change that are being discussed internally. As soon as we define an action, we will let you know. Thanks.

**Comment by Sn-Kinos** [comment]
> @Sn-Kinos sorry that this review is taking longer that normal, but there are some security concerns about this change that are being discussed internally. As soon as we define an action, we will let you know. Thanks.

I understand. I referred this iframe code from another messenger app. I hope it can help you.

**Comment by pvev** [comment]
@Sn-Kinos, thank you for your patience. The decision to use the referrer policy strict-origin-when-cross-origin can expose some sensitive information, which may be considered more or less critical depending on the use case. Therefore, this decision should be made by the workspace administrator.

Given this, the idea is to make this value configurable through a parameter in the admin console. I wanted to ask if you would like to add this new configuration parameter and set the referrer policy based on that parameter (I would help and guide you through this process). Alternatively, I can merge your PR as it is and then take care of adding this configuration parameter in a subsequent PR.

**Comment by Sn-Kinos** [comment]
@pvev I prefer a subsequents PR. I thought there would be no problem because I brought the parameters used in Slack as it is, but I guess it wasn't?

**Review by pvev — APPROVED** [review]
Thank you for the contribution @Sn-Kinos 🚀

**Comment by Sn-Kinos** [comment]
@pvev I fixed lint and type error on CI. How can I re-run actions?

**Comment by Sn-Kinos** [comment]
@pvev I fixed code to pass all tests. Is there anything to do for me to merge this?

**Review by yasserfaraazkhan — APPROVED** [review]
before 
Screenshot 2024-06-11 234834

PR changes are working as expected.

Screenshot 2024-06-11 234851

**Comment by Sn-Kinos** [comment]
Ohhhh I found the case that send shorts link. I'll make it now.

**Comment by Sn-Kinos** [comment]
Come to think of it, it looks good to make a new PR about YouTube Shorts layout. Should I revert the commit?

**Comment by hmhealey** [comment]
> Come to think of it, it looks good to make a new PR about YouTube Shorts layout. Should I revert the commit?

Yeah, if you could revert it and submit it separately, that'd be good. The rest of your changes have gone through reviews and should be all ready to merge, so let's get those merged and add better support for shorts in a separate PR

**Comment by Sn-Kinos** [comment]
> > Come to think of it, it looks good to make a new PR about YouTube Shorts layout. Should I revert the commit?
> 
> Yeah, if you could revert it and submit it separately, that'd be good. The rest of your changes have gone through reviews and should be all ready to merge, so let's get those merged and add better support for shorts in a separate PR

I reverted it! Should I made the issue to make new PR for shorts?

**Comment by hmhealey** [comment]
> I reverted it! Should I made the issue to make new PR for shorts?

Yeah, please do

**Comment by Sn-Kinos** [comment]
I reverted code and passed all tests. Is there anything to do for me to merge this?

**Comment by larkox** [comment]
@esarafianou Friendly reminder on this.

**Comment by larkox** [comment]
@Sn-Kinos Are you up for adding this change? It will involve some (minor) go code changes (to add the config value on the server side). The idea is to add this new setting, some code in the system console (we should decide where exactly, but we are up for suggestions), and only pass these new props if the config allows them. Does that make sense?

If you are not able to do this, we will figure it out (either a new help wanted ticket, or someone does it internally) but may take some time to get this merged.

**Comment by Sn-Kinos** [comment]
@larkox Maybe I can do this. If there are reference, It will be better.

I think this is a bug fix and I was wondering why there should be a configuration for whether or not to provide this feature. Since it is a matter of solving the problem that the original feature is not working, I think that feature should be provided to the user immediately without configuration even if it is delayed a little.

**Comment by pvev** [comment]
> @larkox Maybe I can do this. If there are reference, It will be better.
> 
> I think this is a bug fix and I was wondering why there should be a configuration for whether or not to provide this feature. Since it is a matter of solving the problem that the original feature is not working, I think that feature should be provided to the user immediately without configuration even if it is delayed a little.

@Sn-Kinos thank you for solving this issue, but sharing the referral could be consider sharing sensitive information. That's why we have decided that it should be responsibility of the workspace admin to decide if they want to share the referrer information. We want to set the expose of the referrer behind a config value under `Experimental > Features` as '`YouTube Referrer Policy`'. This would be a boolean value, with the default set to false, meaning the referrer policy remains `no-referrer`, but when set to true the referrer policy would change to `strict-origin-when-cross-origin`. An example of creating the admin config can be found here https://github.com/mattermost/mattermost/pull/22952 . Let me know if you have questions around it.

**Comment by Sn-Kinos** [comment]
> > @larkox Maybe I can do this. If there are reference, It will be better.
> > I think this is a bug fix and I was wondering why there should be a configuration for whether or not to provide this feature. Since it is a matter of solving the problem that the original feature is not working, I think that feature should be provided to the user immediately without configuration even if it is delayed a little.
> 
> @Sn-Kinos thank you for solving this issue, but sharing the referral could be consider sharing sensitive information. That's why we have decided that it should be responsibility of the workspace admin to decide if they want to share the referrer information. We want to set the expose of the referrer behind a config value under `Experimental > Features` as '`YouTube Referrer Policy`'. This would be a boolean value, with the default set to false, meaning the referrer policy remains `no-referrer`, but when set to true the referrer policy would change to `strict-origin-when-cross-origin`. An example of creating the admin config can be found here #22952 . Let me know if you have questions around it.

Thanks! I understand about it. I can work for it. How do I write the description?

**Comment by pvev** [comment]
> > > @larkox Maybe I can do this. If there are reference, It will be better.
> > > I think this is a bug fix and I was wondering why there should be a configuration for whether or not to provide this feature. Since it is a matter of solving the problem that the original feature is not working, I think that feature should be provided to the user immediately without configuration even if it is delayed a little.
> > 
> > 
> > @Sn-Kinos thank you for solving this issue, but sharing the referral could be consider sharing sensitive information. That's why we have decided that it should be responsibility of the workspace admin to decide if they want to share the referrer information. We want to set the expose of the referrer behind a config value under `Experimental > Features` as '`YouTube Referrer Policy`'. This would be a boolean value, with the default set to false, meaning the referrer policy remains `no-referrer`, but when set to true the referrer policy would change to `strict-origin-when-cross-origin`. An example of creating the admin config can be found here #22952 . Let me know if you have questions around it.
> 
> Thanks! I understand about it. I can work for it. How should I write the description?

The description for the system console config value should be something like:

> _Share referrer information when embedding YouTube videos_


And the PR descriptions could be then something like:
_

> This PR introduces a new config for the YouTube referrer policy. The feature allow admins to control whether to expose the referrer information when embedding YouTube videos. By default, the policy is set to false (value is no-referrer) so that no sensitive information is shared. Admins can then change the setting to true (sharing the referrer meaning the value is: strict-origin-when-cross-origin) if they wish to include the referrer information that will unblock the visualization of some youtube videos. The configuration is located under Experimental > Features as 'YouTube Referrer Policy'.

_

cc: @hmhealey

File: webapp/channels/src/components/admin_console/admin_definition.tsx
Comment (larkox) [inline webapp/channels/src/components/admin_console/admin_definition.tsx:6265]: @cwarnermm Can you check these lines?

I guess we should add the final objective of this, which is that this may solve issues with youtube videos previews, but may have some privacy implications.

You may be able to sync with @esarafianou privacy implications.

File: webapp/channels/src/components/youtube_video/youtube_video.test.tsx
Comment (larkox) [inline webapp/channels/src/components/youtube_video/youtube_video.test.tsx:34]: Can we add one test (even if it is a snapshot test) with this setting set to true?

File: webapp/channels/src/components/youtube_video/youtube_video.tsx
Comment (larkox) [inline webapp/channels/src/components/youtube_video/youtube_video.tsx:83]: Apart of the referrerPolicy... The rest of these props are needed? Just for my own knowledge, do you mind explaining what they do and how it differ from the default value?

@esarafianou Just in case you have any concern about any of these other props.

File: webapp/channels/webpack.config.js
Comment (larkox) [inline webapp/channels/webpack.config.js:415]: Just for my own knowledge, what are these changes for?

**Review by larkox — COMMENTED** [review]
Looks great. Nothing really blocking, just a few questions, mostly for my knowledge.

File: webapp/channels/webpack.config.js
Comment (Sn-Kinos) [reply webapp/channels/webpack.config.js:415]: Oh sorry, It's my mistake. That is the config for my local dev env.

File: webapp/channels/src/components/admin_console/admin_definition.tsx
Comment (cwarnermm) [reply webapp/channels/src/components/admin_console/admin_definition.tsx:6265]: @larkox - I propose the following System Console help text update, subject to @esarafianou's input and approval:

"When true, the referrer policy for embedded YouTube videos will be set to "strict-origin-when-cross-origin" which balances the need to protect user privacy with some degree of referral data to support web functionalities, like analytics, logging, and third-party integrations. When false, the referrer policy will be set to "no-referrer" which enhances user privacy by not disclosing the source URL, but limits the ability to track user engagement and traffic sources in analytics tools."

Comment (Sn-Kinos) [reply webapp/channels/src/components/admin_console/admin_definition.tsx:6265]: > @pvev I prefer a subsequents PR. I thought there would be no problem because I brought the parameters used in Slack as it is, but I guess it wasn't?

I just use parameters from Slack. If there are useless parameter and reply it, I'll remove it.

File: webapp/channels/src/components/youtube_video/youtube_video.tsx
Comment (Sn-Kinos) [reply webapp/channels/src/components/youtube_video/youtube_video.tsx:83]: [I just uses parameters from Slack.](https://github.com/mattermost/mattermost/pull/26980#issuecomment-2155201354) If there are useless parameter and reply it, I'll remove it.

File: webapp/channels/src/components/admin_console/admin_definition.tsx
Comment (Sn-Kinos) [reply webapp/channels/src/components/admin_console/admin_definition.tsx:6265]: I think it should contains the content that it can solves the problem that YouTube video preview shows `Video Unavailable`.

Comment (cwarnermm) [reply webapp/channels/src/components/admin_console/admin_definition.tsx:6265]: @Sn-Kinos - which referrer policy results in the ``Video Unavailable`` being resolved?

File: e2e-tests/playwright/support/server/default_config.ts
Comment (pvev) [inline e2e-tests/playwright/support/server/default_config.ts:19]: unnecessary line

File: webapp/channels/src/components/admin_console/admin_definition.tsx
Comment (larkox) [reply webapp/channels/src/components/admin_console/admin_definition.tsx:6265]: @cwarnermm `no-referrer` (what we used to have) can result in Youtube marking the videos as unavailable. The `strict-origin-when-cross-origin` one solves the issue, but as you said, it exposes information.

File: webapp/channels/src/components/youtube_video/youtube_video.tsx
Comment (pvev) [inline webapp/channels/src/components/youtube_video/youtube_video.tsx:81]: shouldn't we better set this directly to 'no-referrer' ?

**Review by pvev — APPROVED** [review]
thank you very much for all the effort on this PR @Sn-Kinos 🚀 🎉 , it looks good to me now. Just two small comments.

File: webapp/channels/src/components/admin_console/admin_definition.tsx
Comment (cwarnermm) [reply webapp/channels/src/components/admin_console/admin_definition.tsx:6265]: @Sn-Kinos - I've updated my [proposed text update](https://github.com/mattermost/mattermost/pull/26980#discussion_r1671176960) above to incorporate your feedback.

File: webapp/channels/src/components/youtube_video/youtube_video.tsx
Comment (Sn-Kinos) [reply webapp/channels/src/components/youtube_video/youtube_video.tsx:81]: I chose to use it indirectly to avoid having to find and change these parts one by one when the `Referrer-Policy` set globally changes. If 'Referrer-Policy' does not change no matter what, I will change it from `undefined` to `no-referrer`

**Comment by yasserfaraazkhan** [comment]
Waiting for CI failures to be fixed.
Will start QA after that.

File: webapp/channels/src/i18n/ko.json
Comment (hmhealey) [inline webapp/channels/src/i18n/ko.json:772]: @Sn-Kinos Would you mind removing these? We appreciate the help translating these, but the translation tool we use runs into conflicts if someone changes the non-English translation files directly

**Review by hmhealey — CHANGES_REQUESTED** [review]

**Comment by Sn-Kinos** [comment]
> Waiting for CI failures to be fixed. Will start QA after that.

Should I update the snapshot of `youtube_video.test.tsx.snap`? I didn't changes it but its snapshot was changed.

image
image

**Comment by pvev** [comment]
> > Waiting for CI failures to be fixed. Will start QA after that.
> 
> Should I update the snapshot of `youtube_video.test.tsx.snap`? I didn't changes it but its snapshot was changed.
> 
> image image

yes please, you can update the snapshots using the `npm run test -u {path}` command. Thanks.

File: webapp/channels/src/components/youtube_video/youtube_video.tsx
Comment (pvev) [reply webapp/channels/src/components/youtube_video/youtube_video.tsx:81]: it shouldn't change. Even for clarity starting it with `no-referrer` works better for me.

Comment (Sn-Kinos) [reply webapp/channels/src/components/youtube_video/youtube_video.tsx:81]: I understood that YouTube should remain 'no-ferrer' even if the global settings change. I'll change it.

**Review by hmhealey — APPROVED** [review]
Thanks for all the work on this. This looks great!

**Comment by hmhealey** [comment]
Just a heads up, I removed an outdated snapshot for a test you added and then removed. Hopefully that should fix CI

**Comment by Sn-Kinos** [comment]
> Just a heads up, I removed an outdated snapshot for a test you added and then removed. Hopefully that should fix CI

I added it according to [this comment](https://github.com/mattermost/mattermost/pull/26980#discussion_r1670009357). Even though I've seen the CLI say it's obsolete, is it okay to remove it?

**Review by pvev — APPROVED** [review]
thank you for this awesome contribution @Sn-Kinos 🚀 , it looks perfect to me now!

**Comment by hmhealey** [comment]
> > Just a heads up, I removed an outdated snapshot for a test you added and then removed. Hopefully that should fix CI
> 
> I added it according to [this comment](https://github.com/mattermost/mattermost/pull/26980#discussion_r1670009357). Even though I've seen the CLI say it's obsolete, is it okay to remove it?

Yeah, I think so. It looks like that's an extra test that you added in https://github.com/mattermost/mattermost/pull/26980/commits/fc7886331f63293bb2b5a9307891cfbfcca76287 which wasn't actually needed. That commit adds only one new test, but it includes snapshots for two different tests. I just removed the second one which didn't have an accompanying test

**Comment by Sn-Kinos** [comment]
It passed all tests but test server creation failed. Is there anything to do for me to solve this problem?

**Comment by Sn-Kinos** [comment]
May I access the test server to test?

**Comment by yasserfaraazkhan** [comment]
> May I access the test server to test?

@Sn-Kinos Yes you can

**Review by yasserfaraazkhan — CHANGES_REQUESTED** [review]
when we play the YT video, it says unavailable.

image

**Comment by Sn-Kinos** [comment]
@yasserfaraazkhan Setting was added by request. I checked it was false and I changed.
image

But it didn't work when I changed the settings. I checked and found a case typo and fixed it.

image

I checked it works with command `MM_SERVICESETTINGS_SITEURL=https://mattermost-pr-26980.test.mattermost.cloud/ npm run dev-server`

**Comment by Sn-Kinos** [comment]
I just fixed typo but CI fails a lot... 😨 What can I do for solve it?

**Comment by larkox** [comment]
@Sn-Kinos Seems like a hiccup in the CI. Let's see if re-running the CI solves it

**Review by yasserfaraazkhan — APPROVED** [review]
able to play the video now.
Thank you @Sn-Kinos

**Comment by Sn-Kinos** [comment]
I'm waiting to develop YouTube Shorts layout. Is it only the Security Review left now?

**Comment by larkox** [comment]
@esarafianou Can you confirm this change?

**Review by esarafianou — APPROVED** [review]
@larkox code looks good but we need someone from Product to agree on adding a new experimental option since we now better track the different product lifecycle paths (experimental - alpha - beta - GA). Adding it as under `ExperimentalSettings` is another thing we need Product to agree on as this makes moving the option out of experimental a breaking change (needing to change the config name).

@wiersgallak  can you let us know what you think?

**Comment by wiersgallak** [comment]
@esarafianou From what I understand, as long as the word "experimental" is not included in the configuration name and it is just under the experimental config setting location, we can move the config setting location without changing the full name of it.

**Comment by Sn-Kinos** [comment]
> @esarafianou From what I understand, as long as the word "experimental" is not included in the configuration name and it is just under the experimental config setting location, we can move the config setting location without changing the full name of it.

Is there anything to do for me to solve it?

**Comment by pvev** [comment]
@Sn-Kinos I think this is ready now to merge. Thank you very much for your contribution and for your patience during our review. Our QE process makes exhaustive reviews to any bit we will put in our code base to always guarantee the best quality and security.

**Pull request merged**
